### PR TITLE
Add Docusaurus

### LINKED ISSUE #58: Design/engine
There are a few issues with our website that I would like to address.

1. Anchor links don't work if you just go by URL, and not clicking on the page. I. e. https://stylelint.io/user-guide/configuration/#extends doesn't scroll to `extends` in Chrome and Safari, but scroll in Firefox. And this problem exists for a very long time. This is some kind of a bug with Phenomic and its dependency [`react-router-scroll`](https://github.com/phenomic/phenomic/blob/0.x/package.json#L69).

2. The design doesn't look good in line with websites of other modern tools.

I'm not sure how to solve the first problem, and it's the most critical, IMHO. I'm afraid it couldn't be solved without a new version of Phenomic. The problem is we use a year old version, and the latest version is in beta for a year. It's [almost ready to be stable](https://github.com/phenomic/phenomic/releases/tag/v1.0.0-beta.3), though. On the other hand, it's a one-person project, and the author could be limited in time to work on it.

So maybe it worth to migrate to another engine, which more people develop.

[Docusaurus](https://docusaurus.io/) is quite popular in open source recently. [Babel](https://babeljs.io/), [Prettier](https://prettier.io/), [Jest](https://jestjs.io/) all use it. One “problem”, that all these websites look similar. But still look good.

[Gatsby](https://www.gatsbyjs.org/) is also popular. It's used for [React website](https://reactjs.org/). Problem with this one, that it has v2 in beta, which doesn't have many docs. So it's not clear is it worth to invest time in v1, when v2 on the horizon. While use v2 without docs will be problematic.

The new engine could help us solve design problems as well. But we could update the design and with old engine as well.

What are your thoughts, @stylelint/contributors? How can we solve scrolling and design problems?

### PR DESCRIPTION
**Pull request by ybiquitous**
Demo: https://stylelint-docusaurus.netlify.com/
(this demo site is updated automatically when commits are pushed)

> Which issue, if any, is this issue related to?

Closes #58

> Is there anything in the PR that needs further explanation?

This PR aims to add [Docusaurus v1](https://docusaurus.io/) as a new doc generator engine.

If you want to try the new website on your computer, please run the following commands (Node 8 required):

<details>
<summary>Commands</summary>

```shell
git clone https://github.com/stylelint/stylelint.io.git

cd stylelint.io

git fetch origin pull/67/head:docusaurus

git checkout docusaurus

npm install

npm run gendoc

cd website

npm install

npm start
```

</details>

Finally, we should remove the current engine [Phenomic](https://phenomic.io/). But I don't want to do so in this PR, because removing it will make it hard for us to review this PR due to too many diffs.
After the migration from Phenomic to Docusaurus in the PR will work, I think it good to remove the old engine (further, update Nodejs and dependencies).

I hope any feedback! 😄 

## About the changes

- Add a new script to generate `.md` files for the website from `node_modules/stylelint` directory: [`scripts/generate-stylelint-docs.js`](https://github.com/stylelint/stylelint.io/pull/67/files#diff-b8d0f3f3ef4afecc31c04e943ddf9448)
  - Run `npm run gendoc` command if you want to run the script.
  - The command outputs the `doc/` directory which Docusaurus recognizes.
  - The script does the followings:
    - Restructure files and directories. E.g., `lib/rules/{rule-name}/README.md` -> `user-guide/rules/{rule-name}.md`
    - Rewrite internal links.
    - Insert [frontmatter metadata](https://docusaurus.io/docs/en/doc-markdown#documents) for Docusaurus.
    - Via [remark](https://github.com/remarkjs/remark).
- Add a new `website` directory which Docusaurus requires. The directory is a core of the website.
- Don't add Docker-related files and `yarn.lock` generated by the [`npx docusaurus-init` command](https://docusaurus.io/docs/en/tutorial-setup#install-the-docusaurus-init-command). Because I think such files may cause new confusions. Let's add them when we will need.

### CODE REVIEW

**Comment by hudochenkov** [comment]
It looks awesome! Keep going, @ybiquitous!

**Comment by ybiquitous** [comment]
Thanks @hudochenkov I'm ready to receive reviews! Welcome feedback!

File: scripts/generate-stylelint-docs.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    mkdir_p(path.dirname(outputFile));
+    fs.writeFileSync(outputFile, output, "utf8");
+    console.log(outputFile);
+  });
+
+  glob.sync("node_modules/stylelint/lib/rules/**/*.md").forEach(file => {
+    const output = processMarkdown(file, {
+      rewriter: url =>
+        url
+          .replace("../indentation/README.md", "indentation.md")
```
Comment (hudochenkov) [inline scripts/generate-stylelint-docs.js]: Why this rule is here? I believe it should be covered by the next `replace()`.

File: website/sidebars.json [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      "user-guide/example-config",
+      "user-guide/css-processors",
+      "user-guide/cli",
+      "user-guide/node-api",
+      "user-guide/postcss-plugin",
+      "user-guide/complementary-tools",
+      "user-guide/articles"
+    ],
+    "Rules": [
+      "user-guide/rules/at-rule-blacklist",
```
Comment (hudochenkov) [inline website/sidebars.json]: It would be nice to generate Rules section automatically if it's possible.

**Comment by ybiquitous** [comment]
> 3. Make “Demo” in the header looks like other links.
> 4. There is no hover state for header links.

I agree to fix the strange style. But this is due to the `siteNavGroupActive` class name by Docusaurus internal logic, it doesn't seem that we can disable this class name… 😕 
image

See `core/nav/HeaderNav.js` in Docusaurus code:

- https://github.com/facebook/docusaurus/blob/e5e085b33d1d9b36a6470847579fd7fee68a12b9/packages/docusaurus-1.x/lib/core/nav/HeaderNav.js#L199
- https://github.com/facebook/docusaurus/blob/e5e085b33d1d9b36a6470847579fd7fee68a12b9/packages/docusaurus-1.x/lib/core/nav/HeaderNav.js#L220-L221

File: scripts/generate-stylelint-docs.js (hunk 1 above)
Comment (ybiquitous) [reply scripts/generate-stylelint-docs.js]: I just forgot about removing it. Thanks. 😅

File: website/sidebars.json (hunk 2 above)
Comment (ybiquitous) [reply website/sidebars.json]: Year, I completely agree 👍 
I will try to find a better solution.

Comment (ybiquitous) [reply website/sidebars.json]: Done.

**Comment by ybiquitous** [comment]
Current progress report:

> 1. Add `html { overflow-y: auto; }` to always have scrollbar visible and avoid header jump, when page is loading.

Done ✅ 

> 2. Remove “Preprocessors” from the header. We want to [deprecate them](https://github.com/stylelint/stylelint/issues/3408), so it's good to not emphasize them in the main navigation.

Done ✅ 

> 5\. On [Demo](https://stylelint-docusaurus.netlify.com/demo) page header is much narrow and links are different (they have hover).

Done ✅

File: website/static/css/custom.css [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+@media only screen and (min-width: 1400px) {
+}
+
+@media only screen and (min-width: 1500px) {
+}
+
+html {
+  overflow-y: auto; /* To always have scrollbar visible and avoid header jump, when page is loading */
```
Comment (hudochenkov) [inline website/static/css/custom.css]: Oh, sorry, I suggested wrong value at first. Could you update, please?

```suggestion
  overflow-y: scroll; /* To always have scrollbar visible and avoid header jump, when page is loading */
```

**Comment by hudochenkov** [comment]
I think we need to revert this change https://github.com/stylelint/stylelint.io/pull/67/commits/626125390cb28c1e9abbea52f5d9a83660b214d7. It's very hacky. Let's find solution for header for Demo (both width and links styles) in separate PR, when this one is merged.

File: website/static/css/custom.css (hunk 3 above)
Comment (ybiquitous) [reply website/static/css/custom.css]: Of course! Done. 👌

**Comment by ybiquitous** [comment]
Yes I agree, it's hacky. 😅 Reverted.

**Comment by ybiquitous** [comment]
> 6\. Is it possible to improve [changelog](https://stylelint-docusaurus.netlify.com/changelog) sidebar. Currently all links looks like “[10.1.0][]”.

Yes, it's strange. When I research it, I find that the behavior is due to the internal code of `docusaurus` and `markdown-toc`. See below:

- https://github.com/facebook/docusaurus/blob/v1.11.1/packages/docusaurus-1.x/lib/core/toc.js#L29
- https://github.com/jonschlinkert/markdown-toc/blob/1.2.0/index.js#L209

Currently, I can't find a solution yet. 😖

**Comment by ybiquitous** [comment]
> 7\. `<meta name="description">` and `<meta property="og:description">` has weird content. On current website they are also incorrect.

It seems to be a Docusaurus's bug. A description of doc pages is generated from the first line of the markdown content. See below:

- https://github.com/facebook/docusaurus/blob/v1.11.1/packages/docusaurus-1.x/lib/core/DocsLayout.js#L92
- Related issue: https://github.com/facebook/docusaurus/issues/1066

File: website/siteConfig.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  onPageNav: "separate",
+
+  scripts: ["/js/pattern-validity.js"],
+
+  algolia: {
+    apiKey: "29d680ce97507c5cd2836c6c74783c05",
+    indexName: "stylelint"
+  },
+
+  gaTrackingId: "UA-72480048-1"
```
Comment (hudochenkov) [inline website/siteConfig.js]: I think we don't need analytics. https://github.com/stylelint/stylelint.io/issues/60 Also having analytics will require to ask users for a consent.

Comment (ybiquitous) [reply website/siteConfig.js]: Oh, sorry. I don't know it. Reverted. 👍

**Comment by ybiquitous** [comment]
> 8\. Is it possible to make different title for a home page? `<title>` and `<meta property="og:title">` is “stylelint · stylelint”. Can we do only one ”stylelint” instead? Or like [Prettier](https://prettier.io/) does: “stylelint · A mighty, modern style linter”?

Nice question. In this PR, the file `docs/index.md` is used for the root page (`/`), which is different from the Prettier case. Prettier has both `/` and `/docs/en/` pages.

However, the title “stylelint · stylelint” is strange. I tried some combinations, [`disableTitleTagline`](https://docusaurus.io/docs/en/site-config#disabletitletagline-boolean) and [`title:`](https://docusaurus.io/docs/en/doc-markdown#documents) frontmatter property, but I couldn't realize “stylelint · A mighty, modern style linter”. 😢 

For details, please see related Docusaurus code:

- https://github.com/facebook/docusaurus/blob/v1.11.1/packages/docusaurus-1.x/lib/core/Site.js#L38-L42
- https://github.com/facebook/docusaurus/blob/v1.11.1/packages/docusaurus-1.x/lib/core/DocsLayout.js#L91

I think the current better solution is having `title: Home` frontmatter property in `docs/index.md`, which outputs the title: “Home · stylelint”. (but `tagline` is never used...)

**Comment by ybiquitous** [comment]
> 9\. Neatpick: Is it possible to configure Autoprefixer (if there any)? I see very old flexbox prefixes in CSS. It would be nice to remove them, if Docusaurus allows it.

Yes, it can. Docusaurus uses Autoprefixer and Browserslist, so we can configure it via `browserslist` field in `package.json` or `browserslistrc`! 👍
Please let me tell a way you prefer! 

See https://github.com/browserslist/browserslist#readme

**Comment by ybiquitous** [comment]
> 10\. Make footer shorter. It take too much space on Demo page.

Done ✅

**Comment by hudochenkov** [comment]
I think the following would be good enough for `browserslist`. Our audience is developers, and developers always have latest versions of browsers :)

```
> 0.5%, last 1 major versions, Firefox ESR, not dead
```

https://browserl.ist/?q=%3E+0.5%25%2C+last+1+major+versions%2C+Firefox+ESR%2C+not+dead

**Comment by hudochenkov** [comment]
I've created issues for unsolved problems, which we could tackle after this PR is merged.

@stylelint/contributors what do you think about new website?

**Comment by evilebottnawi** [comment]
:+1:

**Comment by ybiquitous** [comment]
> 9. Neatpick: Is it possible to configure Autoprefixer (if there any)? I see very old flexbox prefixes in CSS. It would be nice to remove them, if Docusaurus allows it.

Done ✅ 

See https://github.com/stylelint/stylelint.io/pull/67#issuecomment-509772591

**Comment by ybiquitous** [comment]
**NOTE**: I left some notices below.

- Netlify, which is used for this PR demo site, supports auto-redirect. But the current hosting GitHub Pages does NOT support. For example:
  - 🆗 https://stylelint-docusaurus.netlify.com/CHANGELOG -> https://stylelint-docusaurus.netlify.com/changelog
  - 🆖 https://stylelint.io/CHANGELOG -> https://stylelint.io/changelog
  - So I think good if the stylelint new website would be deployed on Netlify. For about this, a discussion seems to be needed.

- I predict the unsolved problems (which have been reported as issues by @hudochenkov, thanks) will NOT be resolved on Docusaurus v1. Because the Docusaurus team is working for v2 release now:
  - The alpha version has been released, current: [`v2.0.0-alpha.19`](https://github.com/facebook/docusaurus/releases/tag/v2.0.0-alpha.19)
  - Related issue: https://github.com/facebook/docusaurus/issues/789
  - The new website has been published: https://v2.docusaurus.io/

- In Docusaurus v2 has some breaking changes, for example, a new config file [`docusaurus.config.js`](https://v2.docusaurus.io/docs/docusaurus.config.js). So I think we will have to rewrite some parts of the PR changes.
  - Sorry, I can't check all breaking changes of Docusaurus v2. 💦

**Comment by hudochenkov** [comment]
Is it possible for Docusaurus to keep file name case? I would like to move to Netlify as well, but migrating engine, hosting, and DNS at the same time is a little bit risky.

**Comment by ybiquitous** [comment]
Sure, Docusaurus outputs `CHANGELOG.html`. When this file is deployed to GitHub Pages, the `/CHANGELOG` access returns 200 status instead of 400, doesn't it?

```
$ cd website

$ npm run build

$ ls build/stylelint/*.html
build/stylelint/CHANGELOG.html       build/stylelint/VISION.html          build/stylelint/developer-guide.html build/stylelint/user-guide.html
build/stylelint/CONTRIBUTING.html    build/stylelint/demo.html            build/stylelint/index.html
```

**Comment by hudochenkov** [comment]
I think I see the problem. Docusaurus generates `CHANGELOG.html`, while Phenomic [`CHANGELOG/index.html`](https://github.com/stylelint/stylelint.github.io/blob/master/CHANGELOG/index.html) so for Github Pages all three links are working:

/CHANGELOG
/CHANGELOG/

But non of this will be working with Github Pages and Docusaurus :(

What we can do is migrate to Netlify with current setup and use https://github.com/stylelint/stylelint.github.io as source for deployment. Then migrate this to Docusaurus completely (with removing Phenomic) and then change Netlify to run on this repository.

**Comment by ybiquitous** [comment]
It looks a good idea to me! 👍

**Comment by ntwb** [comment]
Awesome progress here, thanks a whole bunch @ybiquitous and @hudochenkov for jumping in

**Comment by ybiquitous** [comment]
Thank you for the review!

**Comment by jeddy3** [comment]
Fantastic work!

**Comment by hudochenkov** [comment]
@jeddy3 you're back! Happy to see you :) Even if we don't see each other actually :)

**Comment by jeddy3** [comment]
Thanks! It's nice to be back. It's brilliant to see all this good work on the stylelint website. It's looking great!

**Pull request merged**